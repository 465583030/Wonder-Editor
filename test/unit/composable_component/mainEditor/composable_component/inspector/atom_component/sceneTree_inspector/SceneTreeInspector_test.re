open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe(
    "SceneTreeInspector",
    () => {
      let sandbox = getSandboxDefaultVal();
      beforeEach(
        () => {
          sandbox := createSandbox();
          MainEditorSceneTool.initStateAndGl(~sandbox, ());
          CurrentSourceEditorService.setCurrentSource(EditorType.SceneTree)
          |> StateLogicService.getAndSetEditorState
        }
      );
      afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
      describe(
        "test show component",
        () => {
          test(
            "if hasn't currentGameObject, show nothing",
            () => {
              MainEditorSceneTool.createDefaultScene(
                sandbox,
                MainEditorSceneTool.setFirstBoxTobeCurrentGameObject
              );
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildEmptyAppState(),
                InspectorTool.buildFakeAllShowComponentConfig()
              )
              |> ReactTestTool.createSnapshotAndMatch
            }
          );
          describe(
            "else",
            () => {
              test(
                "if currentGameObject is camera, should show transform and basicCameraView and perspectiveCameraProjection",
                () => {
                  MainEditorSceneTool.createDefaultScene(
                    sandbox,
                    MainEditorSceneTool.setCameraTobeCurrentGameObject
                  );
                  BuildComponentTool.buildInspectorComponent(
                    TestTool.buildEmptyAppState(),
                    InspectorTool.buildFakeAllShowComponentConfig()
                  )
                  |> ReactTestTool.createSnapshotAndMatch
                }
              );
              test(
                "else if currentGameObject is box, should show transform and basicMaterial",
                () => {
                  MainEditorSceneTool.createDefaultScene(
                    sandbox,
                    MainEditorSceneTool.setFirstBoxTobeCurrentGameObject
                  );
                  BuildComponentTool.buildInspectorComponent(
                    TestTool.buildEmptyAppState(),
                    InspectorTool.buildFakeAllShowComponentConfig()
                  )
                  |> ReactTestTool.createSnapshotAndMatch
                }
              )
            }
          );
          describe(
            "test add component workflow",
            () => {
              beforeEach(
                () =>
                  MainEditorSceneTool.createDefaultScene(
                    sandbox,
                    MainEditorSceneTool.setFirstBoxTobeCurrentGameObject
                  )
              );
              test(
                "click the add component button, show addableComponent list",
                () => {
                  let component =
                    BuildComponentTool.buildInspectorComponent(
                      TestTool.buildEmptyAppState(),
                      InspectorTool.buildFakeAllShowComponentConfig()
                    );
                  BaseEventTool.triggerComponentEvent(
                    component,
                    OperateComponentEventTool.triggerClickAddComponentEvent
                  );
                  component |> ReactTestTool.createSnapshotAndMatch
                }
              );
              test(
                "click sourceInstance component, add to inspector",
                () => {
                  let component =
                    BuildComponentTool.buildInspectorComponent(
                      TestTool.buildEmptyAppState(),
                      InspectorTool.buildFakeAllShowComponentConfig()
                    );
                  BaseEventTool.triggerComponentEvent(
                    component,
                    OperateComponentEventTool.triggerClickAddComponentEvent
                  );
                  BaseEventTool.triggerComponentEvent(
                    component,
                    OperateComponentEventTool.triggerClickAddSourceInstanceEvent
                  );
                  let component2 =
                    BuildComponentTool.buildInspectorComponent(
                      TestTool.buildEmptyAppState(),
                      InspectorTool.buildFakeAllShowComponentConfig()
                    );
                  component2 |> ReactTestTool.createSnapshotAndMatch
                }
              )
            }
          )
        }
      );
      describe(
        "deal with specific case",
        () => {
          beforeEach(
            () =>
              MainEditorSceneTool.createDefaultScene(
                sandbox,
                MainEditorSceneTool.setCameraTobeCurrentGameObject
              )
          );
          test(
            "test if specific component not exist, should throw error when parse config from json data",
            () =>
              expect(
                () =>
                  BuildComponentTool.buildInspectorComponent(
                    TestTool.buildEmptyAppState(),
                    InspectorTool.buildFakeErrorAllShowComponentConfig()
                  )
              )
              |> toThrowMessageRe([%re {|/specific\scomponent.+is\serror/mg|}])
          );
          test(
            "test if specific component not exist, should throw error when build component",
            () =>
              expect(
                () =>
                  InspectorTool.buildComponentUIComponent(
                    ("SceneTree", 0),
                    (TestTool.buildEmptyAppState(), TestTool.getDispatch())
                  )
              )
              |> toThrowMessage("the component: SceneTree not exist")
          )
        }
      )
    }
  );