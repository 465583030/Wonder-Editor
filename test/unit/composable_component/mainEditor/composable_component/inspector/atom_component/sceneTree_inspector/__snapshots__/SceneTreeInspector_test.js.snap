// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component else else if currentGameObject is box, should show transform and basicMaterial 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            color : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="#ffffff"
          />
        </article>
      </article>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component else else if currentSceneTreeNode is box, should show transform and basicMaterial 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            color : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="#ffffff"
          />
        </article>
      </article>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component else if currentGameObject is camera, should show transform and basicCameraView and perspectiveCameraProjection 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="40"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicCameraView
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate basic camera view
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          perspectiveCameraProjection
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate perspective camera view
      </div>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component else if currentSceneTreeNode is camera, should show transform and basicCameraView and perspectiveCameraProjection 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="40"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicCameraView
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate basic camera view
      </div>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          perspectiveCameraProjection
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate perspective camera view
      </div>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component if hasn't currentGameObject, show nothing 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            color : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="#ffffff"
          />
        </article>
      </article>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component if hasn't currentSceneTreeNode, show nothing 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            color : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="#ffffff"
          />
        </article>
      </article>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component test add component workflow click sourceInstance component, add to inspector 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            color : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="#ffffff"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          sourceInstance
        </div>
        <span
          className="header-close"
        >
          x
        </span>
      </div>
      <div>
        simulate source instance
      </div>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        add component
      </button>
    </article>
  </article>
</article>
`;

exports[`SceneTreeInspector prepare currentSource is SceneTree test show component test add component workflow click the add component button, show addableComponent list 1`] = `
<article
  className="inspector-component"
>
  <article
    className="inspector-component"
  >
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          transform
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            X : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Y : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Z : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="0"
          />
        </article>
      </article>
    </article>
    <article
      className="componentBox-component"
    >
      <div
        className="header"
      >
        <div
          className="header-triangle"
          onClick={[Function]}
        >
          <span
            className="triangle-bottom"
          />
        </div>
        <div
          className="header-title"
        >
          basicMaterial
        </div>
      </div>
      <article
        className="transform-component"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            color : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="#ffffff"
          />
        </article>
      </article>
    </article>
    <article
      className="addable-component"
    >
      <button
        disabled={false}
        onClick={[Function]}
      >
        add component
      </button>
      <div
        onClick={[Function]}
      >
        sourceInstance
      </div>
    </article>
  </article>
</article>
`;
